### PR TITLE
cmake: fix C++ with newlib

### PR DESCRIPTION
**Pull request by markus-becker-tridonic-com**
When compiling OpenThread within Zephyr and having the following config:

```
CONFIG_NEWLIB_LIBC=y
CONFIG_LIB_CPLUSPLUS=y
```

two problems appear:

1) UINT8_MAX etc. are not defined

```
/home/markus/src/wrp-n4m/modules/lib/openthread/src/ncp/ncp_base_mtd.cpp:
In member function 'otError ot::Ncp::NcpBase::HandlePropertyRemove()
[with unsigned int aKey = 161; otError = otError]':
/home/markus/src/wrp-n4m/modules/lib/openthread/src/ncp/ncp_base_mtd.cpp:898:39:
error: 'UINT8_MAX' was not declared in this scope; did you mean
'UINT_MAX'?
  898 |     VerifyOrExit(serviceDataLength <= UINT8_MAX, error =
OT_ERROR_INVALID_ARGS);
```

2) #include_next <stdint.h> is a warning that gets promoted to an error

```
In file included from
/home/markus/src/wrp-n4m/modules/lib/openthread/src/lib/spinel/spinel.h:316,
                 from
/home/markus/src/wrp-n4m/modules/lib/openthread/src/lib/spinel/spinel.c:46:
/home/markus/src/wrp-n4m/zephyr/lib/libc/newlib/include/stdint.h:28:2:
error: #include_next is a GCC extension
   28 | #include_next <stdint.h>
```

Work-arounds for

1) add `-D__STDC_LIMIT_MACROS` for C++
2) remove `-Werror -pedantic-errors` for C

Are there better fixes?

I am aware that the changes should go to the OpenThread repository, but
for the moment let's discuss here since it is an Zephyr issue right now.

Would it be possible to just patch the CMakeLists.txt?

Signed-off-by: Markus Becker <markus.becker@tridonic.com>

### CODE REVIEW

**Comment by markus-becker-tridonic-com** [comment]
@rlubos please take a look. if this is fine, I can create a PR on zephyr.

**Comment by rlubos** [comment]
Hi @markus-becker-tridonic-com, thanks for reporting the issue. Let me comment on each of the options individually:

1. `-D__STDC_LIMIT_MACROS` - adding option is simple, we do not need to update OT codebase for that. Especially that all of the issues reported here are related to Zephyr itself rather than OpenThread. We have a [zephyr/CmakeLists.txt](https://github.com/zephyrproject-rtos/openthread/tree/zephyr/zephyr) file, which is a glue file for Zephyr and OpenThread build systems. We can easily extend a list of arguments passed to the compiler with something like this:
```
if(CONFIG_NEWLIB_LIBC)
  add_definitions(-D__STDC_LIMIT_MACROS)
endif()
```
I'd see it somewhere around line 90 in the current file.

2. `-Werror` - fortunately this one is already solved in the upstream OpenThread, where they've added a nice switch to enable/disable this option (actually fixed it): https://github.com/openthread/openthread/commit/db765d9370d760489344cfe117124885fbb955c4
All we need is upmerge for this one.

3. `-pedantic-errors` - this one is tricky. I found it impossible to disable this option at our Cmake glue layer. I also find it a bit inconsistent, to allow to disable `-Werror`, while keeping the `-pedantic-errors` option instead of `-pedantic`.  I think that a reasonable approach here would be to introduce a switch to enable/disable pedantic option as well (similar to https://github.com/openthread/openthread/blob/daa626b504b9214fa4477103a062e5e2b4661079/CMakeLists.txt#L42). If such an approach was accepted upstream, we could then upmerge both fixes at once, and fix the remaining issue in our glue layer.

So to summarize, we'd only need to handle `-pedantic-errors` in the upstream OpenThread. Would you mind opening a PR in there?

**Comment by markus-becker-tridonic-com** [comment]
@rlubos upstream has changed.

**Comment by rlubos** [comment]
@markus-becker-tridonic-com Thanks, I'll add OT update to my schedule then.

**Comment by rlubos** [comment]
Superseded by the OpenThread upmerge #11
